[JSC] Support `Promise.all` in async stack trace

sosukesuzuki opened the pull request:
#### 99d0a3203c0295292034106966037bf3f1576c81
<pre>
[JSC] Support `Promise.all` in async stack trace
<a href="https://bugs.webkit.org/show_bug.cgi?id=298600">https://bugs.webkit.org/show_bug.cgi?id=298600</a>

Reviewed by Yusuke Suzuki.

This patch changes async stack trace to support `Promise.all`.

Test: JSTests/stress/async-stack-trace-promise-all-basic.js

Test: JSTests/stress/async-stack-trace-promise-all-basic.js
* JSTests/stress/async-stack-trace-promise-all-basic.js: Added.
(nop):
(shouldThrowAsync):
(throw.new.Error.async fine):
(throw.new.Error.async thrower):
(throw.new.Error.async run):
(async task1):
(async task2):
(async task3):
(async runner):
(throw.new.Error.async drainMicrotasks):
(async foo):
(async bar):
(async baz):
(async drainMicrotasks):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseOnRejectedWithContext):
(linkTimeConstant.promiseAllOnFulfilled):
(linkTimeConstant.promiseAllNewResolveElement):
(all):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsPromiseAllContext):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::promiseAllContextInternalFieldIndex):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseAllContextInternalField):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseAllContextInternalField):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNewInternalFieldObject):
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseAllContextStructure const):
* Source/JavaScriptCore/runtime/JSPromiseAllContext.cpp: Added.
(JSC::JSPromiseAllContext::createWithInitialValues):
(JSC::JSPromiseAllContext::create):
(JSC::JSPromiseAllContext::finishCreation):
(JSC::JSPromiseAllContext::visitChildrenImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromiseAllContext.h: Added.
* Source/JavaScriptCore/runtime/JSPromiseAllContextInlines.h: Added.
(JSC::JSPromiseAllContext::createStructure):
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:

Canonical link: <a href="https://commits.webkit.org/300283@main">https://commits.webkit.org/300283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70e1565d5205938bdbf8ac2b68ad840bfea5e4fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32403 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128594 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74124 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42448 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50327 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92749 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61631 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124983 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33847 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109280 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73405 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32860 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27446 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72088 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/114179 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103354 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131355 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120557 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48970 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37240 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101309 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49344 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105494 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101180 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25650 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46548 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24664 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45684 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48827 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54561 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150715 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48297 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38561 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51647 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49977 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -461,19 +461,48 @@ void Interpreter::getAsyncStackTrace(JSCell* owner, Vector<StackFrame>& results,
-                        return resultGenerator;
+                Structure* reactionsStructure = reactions->structure();
+                unsigned contextFieldAttributes;
+                PropertyOffset contextOffset = reactionsStructure->getConcurrently(vm.propertyNames->builtinNames().contextPrivateName().impl(), contextFieldAttributes);
```
might be worth making an InternalFieldObjectImpl<3> for this. I've seen `get` show up in profiling many times. And this is something that will get called often since the Error stack is materialized in finalizeUnconditionally.

sosukesuzuki replied:
It will affects existing `Promise.all` call, so I'm going to do it after this PR is landed with checking performance regression

Constellation replied:
Hmm, this looks really fragile since it is relying on particular shaped object for the context is only used for Promise.all. I think we should have a better mechanism like, InternalFieldObjectImpl with type etc. for example.

sosukesuzuki replied:
I didn't add any new mechanism, but I simply modified it to use InternalFieldObjectImpl instead. I think this makes it much less fragile now.

Constellation commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -136,7 +136,7 @@ function all(iterable)
     var values = [];
     var globalContext = {
-        promise,
+        @promise: promise,
```
Let's not use private symbols. This object is guaranteed not to be exposed to users. Thus just use normal properties.

sosukesuzuki replied:
Thanks, fixed

Constellation commented on Source/JavaScriptCore/runtime/JSPromiseAllContext.h:
```diff
@@ -0,0 +1,97 @@
+const static uint8_t JSPromiseAllContextNumberOfInternalFields = 4;
+
+class JSPromiseAllContext final : public JSInternalFieldObjectImpl<JSPromiseAllContextNumberOfInternalFields> {
+
```
Remove this empty line.

Constellation commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -151,23 +158,25 @@ function all(iterable)
                 if (constructor !== @Promise)
                     promiseOrCapability = @newPromiseCapabilitySlow(constructor);
-                @performPromiseThen(nextPromise, @promiseAllOnFulfilled, @promiseOnRejectedWithContext, promiseOrCapability, { globalContext, index });
+                var context = @promiseAllContextCreate(promise, values, remainingElementsCountObj, index);
```
Hoist it before this if branch.

Constellation commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -151,23 +158,25 @@ function all(iterable)
                 if (!onRejected)
                     onRejected = @promiseNewOnRejected(promise);
-                then.@call(nextPromise, @promiseAllNewResolveElement(globalContext, index), onRejected);
+                var context = @promiseAllContextCreate(promise, values, remainingElementsCountObj, index);
```
Then, this part is not necessary.

sosukesuzuki committed:
[JSC] Support `Promise.all` in async stack trace
https://bugs.webkit.org/show_bug.cgi?id=298600

Reviewed by Yusuke Suzuki.

This patch changes async stack trace to support `Promise.all`.

Test: JSTests/stress/async-stack-trace-promise-all-basic.js

Test: JSTests/stress/async-stack-trace-promise-all-basic.js
* JSTests/stress/async-stack-trace-promise-all-basic.js: Added.
(nop):
(shouldThrowAsync):
(throw.new.Error.async fine):
(throw.new.Error.async thrower):
(throw.new.Error.async run):
(async task1):
(async task2):
(async task3):
(async runner):
(throw.new.Error.async drainMicrotasks):
(async foo):
(async bar):
(async baz):
(async drainMicrotasks):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseOnRejectedWithContext):
(linkTimeConstant.promiseAllOnFulfilled):
(linkTimeConstant.promiseAllNewResolveElement):
(all):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsPromiseAllContext):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::promiseAllContextInternalFieldIndex):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_getPromiseAllContextInternalField):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_putPromiseAllContextInternalField):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNewInternalFieldObject):
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseAllContextStructure const):
* Source/JavaScriptCore/runtime/JSPromiseAllContext.cpp: Added.
(JSC::JSPromiseAllContext::createWithInitialValues):
(JSC::JSPromiseAllContext::create):
(JSC::JSPromiseAllContext::finishCreation):
(JSC::JSPromiseAllContext::visitChildrenImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromiseAllContext.h: Added.
* Source/JavaScriptCore/runtime/JSPromiseAllContextInlines.h: Added.
(JSC::JSPromiseAllContext::createStructure):
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:

Canonical link: https://commits.webkit.org/300283@main
